sound/soc/sof/virtio-fe.c

```diff
@@ -0,0 +1,734 @@
+		dev_err(sdev->dev, "error: could not send IPC %d\n", ret);
+
+	vfe->msg = msg;
+
+	virtqueue_kick(vfe->ipc_cmd_vq);
+
+	return ret;
+}
+
+/* Handle playback or capture data */
+static void sof_vfe_handle_data(struct virtqueue *vq)
+{
+	struct sof_vfe *vfe = vq->vdev->priv;
+
+	complete(&vfe->completion);
+}
+
+/* Send the IPC message completed. This means vBE has received the cmd */
+static void sof_vfe_cmd_tx_done(struct virtqueue *vq)
+{
+	struct sof_vfe *vfe = vq->vdev->priv;
+	DEFINE_SPINLOCK(lock);
+
+	do {
+		struct snd_sof_ipc_msg *msg = vfe->msg;
+		struct sof_ipc_reply *reply = msg->reply_data;
+		unsigned int len;
+
+		virtqueue_disable_cb(vq);
+
+		spin_lock(&lock);
+		/*
+		 * virtqueue_get_buf() returns the "token" that was provided to
+		 * virtqueue_add_*() functions
+		 */
+		while (virtqueue_get_buf(vq, &len)) {
+			msg->reply_error = reply->error;
+
+			/* Firmware panic? */
+			if (msg->reply_error == -ENODEV)
+				vfe->sdev->ipc->disable_ipc_tx = true;
+
+			msg->ipc_complete = true;
+			wake_up(&msg->waitq);
+		}
+		spin_unlock(&lock);
+	} while (!virtqueue_enable_cb(vq));
+}
+
+static void sof_vfe_posn_update(struct work_struct *work)
+{
+	struct sof_vfe *vfe = container_of(work, struct sof_vfe,
+					   posn_update_work);
+	struct sof_ipc_stream_posn *posn = &vfe->posn;
+	struct virtqueue *vq = vfe->ipc_psn_vq;
+	struct snd_sof_dev *sdev = vfe->sdev;
+	struct snd_sof_pcm *spcm;
+	struct scatterlist sg;
+	unsigned int buflen;
+	int direction;
+
+	/* virtio protects and make sure no re-entry */
+	while (virtqueue_get_buf(vq, &buflen)) {
+		spcm = snd_sof_find_spcm_comp(sdev, posn->comp_id, &direction);
+		if (!spcm) {
+			dev_err(sdev->dev,
+				"err: period elapsed for unused component %d\n",
+					posn->comp_id);
+		} else {
+			/*
+			 * The position update requirement is valid.
+			 * Let's update the position now.
+			 */
+			memcpy(&spcm->stream[direction].posn, posn, sizeof(*posn));
+			snd_pcm_period_elapsed(spcm->stream[direction].substream);
+		}
+
+		/* kick back the empty posn buffer immediately */
+		sg_init_one(&sg, posn, sizeof(*posn));
+		virtqueue_add_inbuf(vq, &sg, 1, posn, GFP_KERNEL);
+		virtqueue_kick(vq);
+	}
+}
+
+/*
+ * handle the pos_update, receive the posn and send to up layer, then
+ * resend the buffer to BE
+ */
+static void sof_vfe_psn_handle_rx(struct virtqueue *vq)
+{
+	struct sof_vfe *vfe = vq->vdev->priv;
+
+	schedule_work(&vfe->posn_update_work);
+}
+
+static int sof_vfe_register(struct snd_sof_dev *sdev)
+{
+	return 0;
+}
+
+static int sof_vfe_unregister(struct snd_sof_dev *sdev)
+{
+	return 0;
+}
+
+#define SOF_VFE_FORMATS (SNDRV_PCM_FMTBIT_S16_LE | SNDRV_PCM_FMTBIT_S24_LE | \
+			 SNDRV_PCM_FMTBIT_S32_LE)
+
+struct snd_soc_dai_driver virtio_dai[] = {
+	{
+		.name = "SSP4 Pin",
+		.playback = SOF_DAI_STREAM("ssp4 Tx", 1, 8,
+				SNDRV_PCM_RATE_8000_192000, SOF_VFE_FORMATS),
+		.capture = SOF_DAI_STREAM("ssp4 Rx", 1, 8,
+				SNDRV_PCM_RATE_8000_192000, SOF_VFE_FORMATS),
+	},
```

Comment:
the guest should not have any notion of Intel-specific DAI hardware.
which is actually a very interesting point since the host should tell you what the DAIs are, and provide the topology. if you only provide the topology to the guest something's not right.

Reply:
guest topology does not contain any physical DAI endpoints, but it contains virtual endpoint connectors for connecting to BE topology components (like mixers, DAIs etc).